---
title: "4-identify_rrna_frags"
format: html
---

```{r}
library(tidyverse)
library(conflicted)
library(Biostrings)
library(patchwork)
library(scales)

conflicts_prefer(dplyr::filter)
```

```{r}
pt <- theme_bw() +
  theme(
    panel.grid = element_blank(),
    legend.background = element_blank(),
    text = element_text(size = 13)
  )
```

Load the depths
```{r}
dlocs <- dir("../alignment/output", pattern = "tsv", full.names = TRUE)

names(dlocs) <- basename(dlocs) %>% 
  str_remove("\\.tsv")

ddf <- lapply(dlocs, read_tsv, col_names = FALSE) %>% 
  bind_rows(.id = "x") %>% 
  separate(x, into = c("org", "enzyme", "repl"), sep = "_")

ddf
```

There's one spot on the 23s rRNA that has way more depth than other positions, the other ones don't contribute that much 
```{r fig.width = 5.5, fig.height = 6.5}
ddf %>% 
  ggplot(., aes(X2, X3, color = repl))+
  geom_line()+
  facet_wrap(~X1, scales = "free", ncol = 1)+
  scale_y_continuous(labels = scales::label_number(scale = 1/1e6, suffix = "M"))+
  scale_color_brewer(name = "Replicate", palette = "Set1")+
  pt+
  theme(legend.text = element_text(size = 10),
        legend.title = element_text(size = 12),
        legend.position = "inside",
        legend.position.inside = c(.18,.93))+
  labs(x = "Position",
       y = "Depth")+
  guides(color = guide_legend(ncol = 2))

ggsave(plot = last_plot(), filename = "../plots/depth_on_rrna.png", width = 5.5, height = 6.5)
```

This is the rrna fasta
```{r}
fa <- readDNAStringSet("../references/ecoli_rrnas.fa")
```

This function takes an rRNA name, two positions to start the graph at (s and e) and two positions to draw vertical lines at (sl and el) to look at and makes graph that shows the positions and the sequence.
```{r}
make_graph <- function(rrna, s, e, sl, el) {
  # get the depths for a specific spot
  depths <- ddf %>%
    dplyr::filter(X1 == rrna & X2 >= s & X2 <= e)
  
  # make the line plot
  p.depths <- depths %>%
    ggplot(., aes(X2, X3, color = repl)) +
    geom_vline(aes(xintercept = sl),
               linetype = 5,
               color = "grey30") +
    geom_vline(aes(xintercept = el),
               linetype = 5,
               color = "grey30") +
    geom_line() +
    facet_wrap(~ X1, scales = "free", ncol = 1) +
    scale_x_continuous(breaks = seq(0, 3000, 1)) +
    scale_y_continuous(labels = scales::label_number(scale = 1 / 1e6, suffix = "M")) +
    scale_color_brewer(name = "Replicate", palette = "Set1") +
    theme_bw() +
    theme(
      panel.grid = element_blank(),
      axis.text.x = element_blank(),
      plot.margin = unit(c(1, 1, 1, 1), "pt"),
      legend.background = element_blank(),
      text = element_text(size = 14),
      legend.text = element_text(size = 10),
      legend.title = element_text(size = 12),
      legend.position = "top",
      legend.direction = "horizontal"
    ) +
    labs(x = NULL, y = "Depth")
  
  # get the sequence
  theseq <- fa[names(fa) == rrna] %>%
    subseq(., start = s, end = e) %>%
    as.character() %>%
    str_split(., "") %>%
    unlist() %>%
    unname()
  
  theseq.df <- data.frame(l = theseq, x = seq(s, e))
  
  p.theseq <- ggplot(theseq.df, aes(x, y = 1, label = l)) +
    geom_text() +
    labs(x = "Position", y = NULL) +
    scale_x_continuous(breaks = seq(0, 2000, 1)) +
    theme_minimal() +
    theme(
      panel.grid = element_blank(),
      axis.text.x = element_text(angle = 90, vjust = 0.5),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      plot.margin = unit(c(1, 1, 1, 1), "pt"),
      text = element_text(size = 14)
    )
  
  final.plot <- (p.depths / p.theseq) + plot_layout(heights = c(.9,.1))
  
  return(final.plot)
}
```

```{r fig.width = 6.5, fig.height = 4}
(seq1 <- make_graph("rna-b2589_23S_ribosomal_RNA", 1505, 1535, 1510, 1531))

ggsave(filename = "../plots/23s_seq1.png", plot = last_plot(), width = 6.5, height = 4)
```

```{r fig.width = 14, fig.height = 4}
(seq2 <- make_graph("rna-b2589_23S_ribosomal_RNA", 1530, 1610, NA, NA))

ggsave(filename = "../plots/23s_seq2.png", plot = last_plot(), width = 14, height = 4)
```

```{r fig.width = 22, fig.height = 4}
make_graph("rna-b3968_16S_ribosomal_RNA", 785, 920, NA, NA)

ggsave(filename = "../plots/16s_seq1.png", plot = last_plot(), width = 22, height = 4)
```

```{r fig.width = 9, fig.height = 4}
make_graph("rna-b3968_16S_ribosomal_RNA", 1035, 1085, 1046, 1074)

ggsave(filename = "../plots/16s_seq1.png", plot = last_plot(), width = 9, height = 4)
```

Presumably then, just counting the reads in the KS137 and 138 samples should show that and/or it's containments being the major species
```{bash}
for file in ../seqdata/4-filtered/*.fq.gz; do
  fname=$(basename $file | cut -d '.' -f 1)
  
  out=../raw_data/dumb_counts/${fname}.txt

  echo "zcat ${file} | awk 'NR%4==2' | sort | uniq -c | sort -nr > ${out}"
done | parallel
```

Those are quite large but it's fine
```{r}
dc.locs <- dir("../raw_data/dumb_counts", full.names = TRUE)

names(dc.locs) <- basename(dc.locs) %>% 
  str_remove("\\.txt")

d.counts <- parallel::mclapply(dc.locs, read_table, col_names = FALSE) %>% 
  bind_rows(.id = "sample")
```

Somehow, the top 7 reads in 137/138 are the same containment of reads, do a quick alignment
```{r}
d.counts %>% 
  group_by(sample) %>% 
  mutate(prop = X1 / sum(X1)*100,
         cuml = cumsum(prop)) %>% 
  ungroup() %>% 
  split(.$sample)
```

```{r}
popular.seqs <- d.counts %>% 
  filter(str_detect(sample, "KS13[78]")) %>% 
  group_by(sample) %>% 
  dplyr::slice(1:7) %>% 
  ungroup() %>% 
  pull(X2) %>% 
  unique()

fa <- DNAStringSet(popular.seqs)

names(fa) <- paste("seq", width(fa), letters[1:length(fa)], sep = "_")

writeXStringSet(fa, "abundant_23s_seqs.fa")
```

```{bash}
clustalo -i abundant_23s_seqs.fa --auto -o 23s.aln --outfmt clu
```

As expected, this one sequence is millions of reads.
```{bash}
cat 23s.aln
```